docs/spec/tuples.rst

```diff
@@ -0,0 +1,155 @@
+can also be used for specializing generic classes or type variables that are
+parameterized using a ``TypeVarTuple``. For more details, see
+:ref:`args_as_typevartuple`.
+
+
+Type Compatibility Rules
+------------------------
+
+Because tuple contents are immutable, the element types of a tuple are covariant.
+For example, ``tuple[int, int]`` is a subtype of ``tuple[float, complex]``.
+
+As discussed above, a homogeneous tuple of arbitrary length is equivalent
+to a union of tuples of different lengths. That means ``tuple[()]``, 
+``tuple[int]`` and ``tuple[int, *tuple[int, ...]]`` are all subtypes of
+``tuple[int, ...]``. The converse is not true; ``tuple[int, ...]``` is not a
+subtype of ``tuple[int]``.
+
+The type ``tuple[Any, ...]`` is bidirectionally compatible with any tuple::
+
+    def func(t1: tuple[int], t2: tuple[int, ...], t3: tuple[Any, ...]):
+        v1: tuple[int, ...] = t1  # OK
+        v2: tuple[Any, ...] = t1  # OK
+
+        v3: tuple[int] = t2  # Type error
+        v4: tuple[Any, ...] = t2  # OK
+
+        v5: tuple[float, float] = t3  # OK
+        v6: tuple[int, *tuple[str, ...]] = t3  # OK
+
+
+The length of a tuple at runtime is immutable, so it is safe for type checkers
+to use length checks to narrow the type of a tuple::
+
+    def func(val: tuple[int] | tuple[str, str] | tuple[int, *tuple[str], int]):
+        if len(val) == 1:
+            # Type can be narrowed to tuple[int].
+            reveal_type(val)  # tuple[int]
+        
+        if len(val) == 2:
+            # Type can be narrowed to tuple[str, str] | tuple[int, int].
+            reveal_type(val)  # tuple[str, str] | tuple[int, int]
+
+        if len(val) == 3:
+            # Type can be narrowed to tuple[int, str, int].
+            reveal_type(val)  # tuple[int, str, int]
+
+This property may also be used to safely narrow tuple types within a ``match``
+statement that uses sequence patterns.
+
+If a tuple element is a union type, the tuple can be safely expanded into a
+union of tuples. For example, ``tuple[int | str]`` is equivalent to 
+``tuple[int] | tuple[str]``. If multiple elements are union types, the expansion
+must consider all combinations. For example, ``tuple[int | str, int | str]`` is
+equivalent to ``tuple[int, int] | tuple[int, str] | tuple[str, int] | tuple[str, str]``.
+Unbounded tuples cannot be expanded in this manner. 
+
+Type checkers may safely use this equivalency rule when narrowing tuple types::
+
+    def func(subj: tuple[int | str, int | str]):
+        match subj:
+            case x, str():
+                reveal_type(subj)  # tuple[int | str, str]
+            case y:
+                reveal_type(subj)  # tuple[int | str, int]
+
+The ``tuple`` class derives from ``Sequence[T_co]`` where ``T_co`` is a covariant
+(non-variadic) type variable. The specialized type of ``T_co`` should be computed
+by a type checker based on the union of element types within the tuple.
```

Comment:
Does this mean that given

```
x: Any
y: int
```

the type of `(x, y)` is `Sequence[Any | int]` and not `Sequence[Any]`?

Reply:
> the type of (x, y) is Sequence[Any | int] and not Sequence[Any]?

Yes, the type would be `Sequence[Any | int]` based on the current draft text.

It looks like mypy uses a join operation here rather than a union, so it produces `Sequence[Any]` rather than `Sequence[Any | int]`. Maybe the spec needs to be less prescriptive here to allow for both unions and joins (or any other operation that produces some type that is a supertype of all element types). I'll post that question in the discussion to get feedback from the broader community.

(There are many reasons why a union is better than a join here. There are [63 active bug reports](https://github.com/python/mypy/labels/topic-join-v-union) in the mypy issue tracker that are labeled "topic-join-v-union".)